### DISCUSSION TITLE
How can we handle when there are more services and other 3rd party dependencies between them.

### DISCUSSION THREAD

**Discussion by joobisb** [MarioCarrion/todo-api-microservice-example#319] [Q&A] [answered]
Let's say there are `task`, `user`, `subtask` services, we need to send notifications when we create a `task`or `subtask`, this notification can be via different channels email, SMS, etc. We may use different 3rd party API providers for each of them, etc. How would we structure this ?

Also another question, i.e whenever we need to add a new service it should be independent of each other right? What happens when a service wants to use another service ? how do we communicate between these services or should we duplicate the functionality in that case for each service ?

**Comment by MarioCarrion** [comment, marked as answer]
@joobisb

> Let's say there are task, user, subtask services, we need to send notifications when we create a taskor subtask, this notification can be via different channels email, SMS, etc. We may use different 3rd party API providers for each of them, etc. How would we structure this ?

It sounds like like you mean customer notifications after the fact when we know we already processed everything we needed to process, in that case I'd use a queue with a payload that indicates:
* `notificationType` to use such as email, SMS, etc,
* `notificationProvider` like the actual 3rd party provider, maybe you have two different for email, for example, and
* `message` which will indicate the content itself.
 
For the code I'd probably organize those in a new package `internal/notifications` and do something like what I did with the other packages that represent data stores in `internal`.

> Also another question, i.e whenever we need to add a new service it should be independent of each other right? What happens when a service wants to use another service ? how do we communicate between these services or should we duplicate the functionality in that case for each service ?

Service is such overloaded term, I need a bit of clarification; do you mean service as in `service/task.go` or "third party service"?

**Reply by joobisb** [reply]
@MarioCarrion Thanks for answering this 

> Service is such overloaded term, I need a bit of clarification; do you mean service as in service/task.go or "third party service"?

I meant service as in service/task.go

**Reply by MarioCarrion** [reply]
@joobisb Typically you want to have independent service types; _but_ if you have the need to call service `A` from service `B`; then `B` would receive `A` as argument and orchestrate the calls internally.

Another approach (if both share the same data store) is to _combine them_ to match your domain, that way the transactions are handled correctly.

In my experience what drivers either choice (or a mix of both) is how those service types are used, for example how the final user interacts with them in the application you're building.
